nixos: Add `system.build.`{`toplevel`,`installBootLoader`}, improve error message

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

 - Undocumented options are bad, even more so when they're important ones like `toplevel`
 - [Bad error message](https://github.com/NixOS/nixpkgs/issues/156373)

This is the successor of https://github.com/NixOS/nixpkgs/pull/155883

Things done:
 - add temporary helper function
 - add `types.unique { message } type`
 - make `system.build.toplevel` non-internal
 - use `types.unique` to improve error message:

Original error:

```
building the system configuration...
error: The option `system.build.installBootLoader' is defined multiple times.
       Definition values:
       - In `nixos/modules/system/boot/loader/grub/grub.nix': <derivation /nix/store/v1yj7wffg4q1mqk8rahsq5pgvq1jagn8-install-grub.sh.drv>
       - In `nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix': <derivation /nix/store/kzmd23ma820f3kxx6sylnkc2yfa348cm-systemd-boot.drv>
```

New error:

```
nix-repl> (nixos { fileSystems."/".device = "x"; boot.loader.grub.devices = ["/dev/x"]; boot.loader.systemd-boot.enable = true; boot.loader.grub.enable = true; }).config.system.build.installBootLoader
error: The option `system.build.installBootLoader' is defined multiple times.
       Only one bootloader can be enabled at a time. This requirement has not
       been checked until NixOS 22.05. Earlier versions defaulted to the last
       definition. Change your configuration to enable only one bootloader.

       Definition values:
       - In `nixos/modules/system/boot/loader/grub/grub.nix': <derivation /nix/store/v1yj7wffg4q1mqk8rahsq5pgvq1jagn8-install-grub.sh.drv>
       - In `nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix': <derivation /nix/store/kzmd23ma820f3kxx6sylnkc2yfa348cm-systemd-boot.drv>
```





###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable: 
  - [x]  **manually checked nixosTests.nginx drv hash unchanged, manually tested the error** 
  - [ ] automated
    - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
    - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
    - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

roberth committed:
nixos/system/build: Extract

Modules that do not depend on e.g. toplevel should not have to include it just to set
things in `system.build`. As a general rule, this keeps tests simple, usage flexible
and evaluation fast. While one module is insignificant, consistency and good practices
are.

roberth committed:
nixos: Make system.build a submodule with freeformType

This allows the values below it to be specified as options, while
remaining compatible with existing code.

roberth committed:
nixos: Document system.build.toplevel

roberth committed:
lib.types: Add unique like uniq, but custom errors

Couldn't extend types.uniq and it had a silly name anyway.
Now we can have better error messages.

roberth committed:
nixos: Make system.build.installBootLoader a proper option

This improves the error message when the configuration contains
more than one boot loader.

ncfavier commented on lib/types.nix:
```diff
@@ -470,6 +470,18 @@ rec {
       nestedTypes.elemType = elemType;
     };
 
+    unique = { message }: type: mkOptionType rec {
```
Same question: shouldn't this replace `uniq`?

roberth replied:
That would be an unnecessary breaking change for out-of-tree usage of the module system. `uniq` is an awful name anyway, so I'd rather just deprecate that at some point.

ncfavier replied:
I meant something like `uniq = unique { message = "" };`, maybe with a deprecation warning. Sorry, that wasn't clear

roberth replied:
I think it's ok not to deprecate or not deprecate yet.

roberth commented on nixos/modules/system/activation/top-level.nix:
```diff
@@ -139,21 +137,13 @@ in
 
 {
   imports = [
+    ../build.nix
```
That would work, but only as long as `top-level` isn't used independently of `module-list.nix`.
Other than that it makes no practical difference.

roberth committed:
nixos: Explain system.build.installBootLoader's odd default

I don't really approve of this solution, but documenting its
purpose was the least I could do for now.

roberth commented on nixos/modules/system/activation/top-level.nix:
```diff
@@ -224,6 +214,36 @@ in
+    system.build = lib.tmpFixupSubmoduleBoundary {
+      installBootLoader = mkOption {
+        internal = true;
+        default = "echo 'Warning: do not know how to make this configuration bootable; please enable a boot loader.' 1>&2; true";
```
I've added a comment:
```
        # "; true" => make the `$out` argument from switch-to-configuration.pl
        #             go to `true` instead of `echo`, hiding the useless path
        #             from the log.
        default = "echo 'Warning: do not know how to make this configuration bootable; please enable a boot loader.' 1>&2; true";
```

ncfavier replied:
Ah, thanks, didn't realise it was called with `$out`

roberth commented on lib/options.nix:
```diff
@@ -134,6 +134,12 @@ rec {
       throw "The unique option `${showOption loc}' is defined multiple times. Definition values:${showDefs defs}"
     else (head defs).value;
 
+  mergeUniqueOption = { message }: loc: defs:
```
`mergeOneOption` is a public `lib` function with a bad name.

I'm not making the message optional, because it seems like a good idea to require it.

ncfavier replied:
Same as below, I think we could at least do `mergeOneOption = mergeUniqueOption { message = ""; }` to avoid duplicated code.

Also not sure I agree about requiring a more detailed error message. Why here in particular?

ncfavier replied:
For example what should the message be when merging several values of `package` type? I think the base message is self-explanatory

roberth replied:
Fair enough, perhaps many options are quite obvious and don't need it.

roberth commented on lib/options.nix:
```diff
@@ -283,4 +289,13 @@ rec {
 
   unknownModule = "<unknown-file>";
 
+  /* Workaround for https://github.com/NixOS/nixpkgs/issues/146882
+     This will be removed without notice; thanks for understanding.
+  */
+  tmpFixupSubmoduleBoundary = subopts:
```
The argument isn't a submodule though.

The evocative name is intentional, so nobody uses it externally and we can use a find and replace when implemented properly.

The alternative is to just inline its definition and call it a day, but I would prefer not to.

roberth replied:
Not inlined but moved to a local let binding, achieving the same goal of making this non-public and non-reusable.

ncfavier commented:
@ofborg test grub systemd-boot.basic

roberth committed:
lib.modules: Define mergeOneOption in terms of mergeUniqueOption

roberth committed:
nixos/doc: Document types.unique

lopsided98 commented:
Previously it was only possible to override an attr in `system.build` using `mkAfter`:
```
system.build = mkAfter {
  attr = ...;
};
```

This now fails with `lib/types.nix:569:38 called with unexpected argument 'priority'`. The correct way is now to use `mkForce` like any other option.

I don't think this breakage needs to be fixed, since the new config is definitely better than the old `mkAfter` hack. I'm just posting this to help anyone else who runs into this cryptic error message.
